`throwOnParseFailures` Option Not Honored

Sometimes, formatting mistakes in files cause us to get the following error:
```
There were problems parsing some source files, run with --info to see full stack traces
```
Why they surface as "parse problems" instead of normal formatting mistakes is unknown to me, but it's good enough! : v)

The issue is that we rely on CI to check our formatting with:
```
./gradlew rewriteDryRun --no-parallel --no-configuration-cache
```
And unlike regular formatting mistakes, these errors don't cause our CI to fail, meaning formatting mistakes go undetected.

So fast-forward to today, and I found the amazing, exactly-what-I-was-looking-for `throwOnParseFailures` option!
(At the bottom of https://github.com/openrewrite/rewrite-gradle-plugin/blob/main/docs/plugin-reference.md)
Buttt, setting it to `true` didn't make a difference.

So, like any sane person, I cloned the repo and grepped for `throwOnParseFailures`, and it doesn't look like the option is actually checked anywhere? It's defined, has getters/setters, but never called or used that I can see.

If I'm missing something, feel free to correct me! : v)
I'm also more than happy to poke around the history and open a PR to implement `throwOnParseFailures`,
but wanted to make sure I wasn't missing something obvious before doing so!

----

## What version of OpenRewrite are you using?

At risk of embarrassing myself on the internet, I don't know how to print the plugin version through gradle ;-;

## How are you running OpenRewrite?
We're using the gradle plugin:
`apply plugin: 'org.openrewrite.rewrite'`

With the following configuration:
```
rewrite {
    configFile = "${rootProject.projectDir}/../../../rewrite.yml"

    activeRecipe('com.zeroc.IceRewriteRecipes')
    activeStyle('com.zeroc.IceRewriteStyle')

    throwOnParseFailures = true
    exportDatatables = true

    exclusion(
        // OpenRewrite is technically capable of modifying Groovy and Kotlin files, but we are avoiding this for now.
        '**/*.gradle',
        '**/*.kts',
        // These are all generated files, and thus should not be subject to modification from OpenRewrite.
        '**/generated/**/*',
    )
}
```
Our project is public, and we have a tracking issue about this!
- Please see: https://github.com/zeroc-ice/ice/issues/4732

## What is the smallest, simplest way to reproduce the problem?
To be honest, I'm not entirely sure! There's an android test file we have, and certain formatting errors cause
```
> Task :rewriteDryRun
Validating active recipes
Scanning sources in project :
Using active styles [com.zeroc.IceRewriteStyle, org.openrewrite.java.Checkstyle]
There were problems parsing some source files, run with --info to see full stack traces
There were problems parsing java\test\android\controller\src\main\java\com\zeroc\testcontroller\ControllerActivity.java
All sources parsed, running active recipes: com.zeroc.IceRewriteRecipes
Applying recipes would make no changes. No report generated.

[Incubating] Problems report is available at: file:///D:/Code/Workspace/ice/java/test/android/controller/build/reports/problems/problems-report.html
```

## What did you expect to see?
I saw what I expected, but I also expected it to fail my build, and instead it succeeds with:
```
BUILD SUCCESSFUL in 6s
5 actionable tasks: 1 executed, 4 up-to-date
```

## What did you see instead?
```
BUILD FAILED
```
would be preferable : v)

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?

I'd be totally happy to contribute a fix, maybe with some gentle guidance : v)
I'm very familiar with Java, but not very familiar with the inner workings of this project.

--- a/plugin/src/main/java/org/openrewrite/gradle/isolated/DefaultProjectParser.java
+++ b/plugin/src/main/java/org/openrewrite/gradle/isolated/DefaultProjectParser.java
@@ -1428,6 +1428,9 @@ public class DefaultProjectParser implements GradleProjectParser {
                 })
                 .peek(xmlDetector::sample)
                 .collect(toList());
+        if (extension.getThrowOnParseFailures() && firstWarningLogged.get()) {
+            throw new RuntimeException("There were problems parsing some source files, run with --info to see full stack traces. Fix the parse failures or set throwOnParseFailures to false to continue.");
+        }
         Map<Class<? extends SourceFile>, NamedStyles> stylesByType = new HashMap<>();
         stylesByType.put(J.CompilationUnit.class, javaDetector.build());
         stylesByType.put(K.CompilationUnit.class, kotlinDetector.build());

--- a/plugin/src/test/kotlin/org/openrewrite/gradle/GradleRunnerTest.kt
+++ b/plugin/src/test/kotlin/org/openrewrite/gradle/GradleRunnerTest.kt
@@ -25,6 +25,14 @@ val gradleVersion: String? = System.getProperty("org.openrewrite.test.gradleVers
 interface GradleRunnerTest {
 
     fun runGradle(testDir: File, vararg args: String): BuildResult {
+        return gradleRunner(testDir, *args).build()
+    }
+
+    fun runGradleAndFail(testDir: File, vararg args: String): BuildResult {
+        return gradleRunner(testDir, *args).buildAndFail()
+    }
+
+    private fun gradleRunner(testDir: File, vararg args: String): GradleRunner {
         return GradleRunner.create()
             .withDebug(ManagementFactory.getRuntimeMXBean().inputArguments.toString().indexOf("-agentlib:jdwp") > 0)
             .withProjectDir(testDir)
@@ -36,6 +44,5 @@ interface GradleRunnerTest {
             .withArguments(*args, "--info", "--stacktrace")
             .withPluginClasspath()
             .forwardOutput()
-            .build()
     }
 }

--- a/plugin/src/test/kotlin/org/openrewrite/gradle/RewriteDryRunTest.kt
+++ b/plugin/src/test/kotlin/org/openrewrite/gradle/RewriteDryRunTest.kt
@@ -348,6 +348,43 @@ class RewriteDryRunTest : RewritePluginTest {
             )
     }
 
+    @Issue("https://github.com/openrewrite/rewrite-gradle-plugin/issues/441")
+    @Test
+    fun `throwOnParseFailures fails the build when parsing fails`() {
+        gradleProject(projectDir) {
+            buildGradle(
+                """
+                plugins {
+                    id("org.openrewrite.rewrite")
+                }
+
+                repositories {
+                    mavenLocal()
+                    mavenCentral()
+                    maven {
+                       url = uri("https://central.sonatype.com/repository/maven-snapshots")
+                    }
+                }
+
+                rewrite {
+                    activeRecipe("org.openrewrite.FindSourceFiles")
+                    throwOnParseFailures = true
+                }
+                """
+            )
+            textFile(
+                "broken.yml",
+                """
+                key: value
+                  - invalid indentation
+                    foo: [unclosed
+                """.trimIndent()
+            )
+        }
+        val result = runGradleAndFail(projectDir, taskName())
+        assertThat(result.output).contains("Fix the parse failures or set throwOnParseFailures to false to continue")
+    }
+
     @EnabledIfEnvironmentVariable(named = "ANDROID_HOME", matches = ".*")
     @EnabledForGradleRange(min = "5.0", max = "7.6.4")
     @Test

